src/charm.py

```diff
@@ -115,6 +116,33 @@ def _on_install(self, _) -> None:
 
         self.unit.status = WaitingStatus("Waiting for relations")
 
+    def _on_upgrade_charm(self, _) -> None:
+        """Update the mysql-router config on charm upgrade."""
+        if isinstance(self.unit.status, ActiveStatus):
+            self.unit.status = MaintenanceStatus("Upgrading charm")
+
+            requires_data = json.loads(self.app_peer_data.get(MYSQL_ROUTER_REQUIRES_DATA))
+            related_app_name = (
+                self.shared_db_relation._get_related_app_name()
+                if self.shared_db_relation._shared_db_relation_exists()
+                else self.database_provides_relation._get_related_app_name()
+            )
+
+            try:
+                MySQLRouter.bootstrap_and_start_mysql_router(
+                    requires_data["username"],
+                    self._get_secret("app", "database-password"),
+                    related_app_name,
+                    requires_data["endpoints"].split(",")[0].split(":")[0],
+                    "3306",
+                    True,
+                )
+            except MySQLRouterBootstrapError:
+                self.unit.status = BlockedStatus("Failed to bootstrap mysqlrouter")
+                return
+
+            self.unit.status = ActiveStatus()
+
```

Comment:
Not strictly necessary, but I think we should be reconfiguring running charms. Is this OK to add?

Reply:
I think that it should be OK to add - although, reconfiguring running charms is territory we havent explored in any mysql charms

Reply:
It will regenerate the config file and I think keep a backup of the old one. Without it the bootstrap will complain that configuration is already present.